.github/workflows/test-pr-publish.yml

```diff
@@ -0,0 +1,49 @@
+name: NPM PR Publish
+
+on:
+  issue_comment:
+    types: [created]
+
+jobs:
+  pr-commented:
+    # This job only runs for pull request comments
+    name: Publish PR
+
+    if: ${{ github.event.issue.pull_request && contains(github.event.comment.body, 'Publish PR version') }}
+
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/github-script@v3
+        id: get-pr
+        with:
+          script: |
+            const request = {
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              pull_number: context.issue.number
+            }
+            core.info(`Getting PR #${request.pull_number} from ${request.owner}/${request.repo}`)
+            try {
+              const result = await github.pulls.get(request)
+              return result.data
+            } catch (err) {
+              core.setFailed(`Request failed with error ${err}`)
+            }
```

Comment:
Is the reason you're going through all this trouble to avoid publishing a lot of versions? I think it might be worth it to try and optimize in a different way cause this is way more complex than just running publish on a `pull_request` event

For instance, one thought might be to introduce a new "Publish PR Version" label and use the pull_request event hook with type of `labeled` -- https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request

![image](https://user-images.githubusercontent.com/1666298/111379678-c89f9b00-8679-11eb-87eb-e572f2b0b410.png)

```yml
on:
  pull_request:
    types: [labeled]
```

then have an step or `if` that prevents it from running unless the label is on the PR. that way you wouldnt need to run any scripts, no fetching, no setting outputs cause you have everything from the `github` object already

```yml
on:
  pull_request:
    types: [labeled]

jobs:
  publish-pr:
    runs-on: ubuntu-latest

    # you'll need to test this for sure!
    if: github.event.label.name == 'pr-version'

    steps:
      - uses: actions/checkout@v2

      - uses: actions/setup-node@v1
        with:
          node-version: 12.x

      - name: publish to npm
        run: |
          npm version "0.0.0-PR-${PR}--${SHA}"
          npm publish --access public --tag pr --dry-run
        env:
          PR: ${{ github.event.number }}
          SHA: ${{ github.event.pull_request.head.sha }}
          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
```

Reply:
That's a really interesting alternative - and yes, that would certainly simplify things considerably. The main reason is when commenting on a PR, it's an `issue_comment`, and not a pull request event. This means we don't get the PR context we otherwise would with a PR event. It requires fetching the necessary data via the API with the info obtained in the github-script step.
I just liked the idea of commenting to trigger a publish action, but toggling a label could work..